// ng-date-format/bundles/ng-date-format.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ng-date-format', ['exports', '@angular/core', '@angular/common'], factory) :
    (factory((global['ng-date-format'] = {}),global.ng.core,global.ng.common));
}(this, (function (exports,core,common) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var DatePipe = (function () {
        function DatePipe(datePipe) {
            this.datePipe = datePipe;
        }
        /**
         * @param {?} value
         * @param {?=} args
         * @return {?}
         */
        DatePipe.prototype.transform = /**
         * @param {?} value
         * @param {?=} args
         * @return {?}
         */
            function (value, args) {
                /** @type {?} */
                var countDigits = 0;
                /** @type {?} */
                var dateStr = '';
                /** @type {?} */
                var defaultFormat = 'mediumDate';
                if (args) {
                    defaultFormat = args;
                }
                if ((value === null) || (value === undefined)) {
                    return value;
                }
                if (!isNaN(value)) {
                    countDigits = value.toString().length;
                    if (countDigits === 13) {
                        /** @type {?} */
                        var dtObj = new Date(value);
                        dateStr = this.datePipe.transform(dtObj, defaultFormat);
                    }
                    else if (countDigits === 10) {
                        /** @type {?} */
                        var dtObj = new Date(value * 1000);
                        dateStr = this.datePipe.transform(dtObj, defaultFormat);
                    }
                }
                else {
                    /** @type {?} */
                    var dtObj = new Date(value);
                    dateStr = this.datePipe.transform(dtObj, defaultFormat);
                }
                return dateStr;
            };
        DatePipe.decorators = [
            { type: core.Pipe, args: [{
                        name: 'date'
                    },] },
        ];
        /** @nocollapse */
        DatePipe.ctorParameters = function () {
            return [
                { type: common.DatePipe }
            ];
        };
        return DatePipe;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var DateFormatModule = (function () {
        function DateFormatModule() {
        }
        DateFormatModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [
                            common.CommonModule
                        ],
                        declarations: [DatePipe],
                        exports: [
                            DatePipe
                        ],
                        providers: [
                            common.DatePipe
                        ]
                    },] },
        ];
        return DateFormatModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    exports.DatePipe = DatePipe;
    exports.DateFormatModule = DateFormatModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
