// ng-date-format/fesm2015/ng-date-format.js
import { Pipe, NgModule } from '@angular/core';
import { DatePipe, CommonModule } from '@angular/common';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class DatePipe$1 {
    /**
     * @param {?} datePipe
     */
    constructor(datePipe) {
        this.datePipe = datePipe;
    }
    /**
     * @param {?} value
     * @param {?=} args
     * @return {?}
     */
    transform(value, args) {
        /** @type {?} */
        let countDigits = 0;
        /** @type {?} */
        let dateStr = '';
        /** @type {?} */
        let defaultFormat = 'mediumDate';
        if (args) {
            defaultFormat = args;
        }
        if ((value === null) || (value === undefined)) {
            return value;
        }
        if (!isNaN(value)) {
            countDigits = value.toString().length;
            if (countDigits === 13) {
                /** @type {?} */
                const dtObj = new Date(value);
                dateStr = this.datePipe.transform(dtObj, defaultFormat);
            }
            else if (countDigits === 10) {
                /** @type {?} */
                const dtObj = new Date(value * 1000);
                dateStr = this.datePipe.transform(dtObj, defaultFormat);
            }
        }
        else {
            /** @type {?} */
            const dtObj = new Date(value);
            dateStr = this.datePipe.transform(dtObj, defaultFormat);
        }
        return dateStr;
    }
}
DatePipe$1.decorators = [
    { type: Pipe, args: [{
                name: 'date'
            },] },
];
/** @nocollapse */
DatePipe$1.ctorParameters = () => [
    { type: DatePipe }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class DateFormatModule {
}
DateFormatModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule
                ],
                declarations: [DatePipe$1],
                exports: [
                    DatePipe$1
                ],
                providers: [
                    DatePipe
                ]
            },] },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

export { DatePipe$1 as DatePipe, DateFormatModule };
