// ngx-malihu-scrollbar/esm5/src/malihu-scrollbar.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable, NgZone } from '@angular/core';
var MalihuScrollbarService = /** @class */ (function () {
    function MalihuScrollbarService(zone) {
        this.zone = zone;
    }
    /**
     * @param {?} element
     * @param {?} options
     * @return {?}
     */
    MalihuScrollbarService.prototype.initScrollbar = /**
     * @param {?} element
     * @param {?} options
     * @return {?}
     */
    function (element, options) {
        /** @type {?} */
        var jQueryElement = this.getElement(element);
        this.zone.runOutsideAngular(function () { return jQueryElement.mCustomScrollbar(options); });
        if (jQueryElement.length > 0 && jQueryElement[0].tagName === 'BODY') {
            jQueryElement[0].style.position = 'absolute';
            jQueryElement[0].style.overflow = 'auto';
            jQueryElement[0].style.height = '100vh';
            jQueryElement[0].style.width = '100vw';
        }
    };
    /**
     * @param {?} element
     * @param {?} parameter
     * @param {?} options
     * @return {?}
     */
    MalihuScrollbarService.prototype.scrollTo = /**
     * @param {?} element
     * @param {?} parameter
     * @param {?} options
     * @return {?}
     */
    function (element, parameter, options) {
        var _this = this;
        this.zone.runOutsideAngular(function () { return _this.getElement(element).mCustomScrollbar('scrollTo', parameter, options); });
    };
    /**
     * @param {?} element
     * @return {?}
     */
    MalihuScrollbarService.prototype.update = /**
     * @param {?} element
     * @return {?}
     */
    function (element) {
        var _this = this;
        this.zone.runOutsideAngular(function () { return _this.getElement(element).mCustomScrollbar('update'); });
    };
    /**
     * @param {?} element
     * @return {?}
     */
    MalihuScrollbarService.prototype.stop = /**
     * @param {?} element
     * @return {?}
     */
    function (element) {
        var _this = this;
        this.zone.runOutsideAngular(function () { return _this.getElement(element).mCustomScrollbar('stop'); });
    };
    /**
     * @param {?} element
     * @param {?=} reset
     * @return {?}
     */
    MalihuScrollbarService.prototype.disable = /**
     * @param {?} element
     * @param {?=} reset
     * @return {?}
     */
    function (element, reset) {
        var _this = this;
        this.zone.runOutsideAngular(function () { return _this.getElement(element).mCustomScrollbar('disable', !!reset); });
    };
    /**
     * @param {?} element
     * @return {?}
     */
    MalihuScrollbarService.prototype.destroy = /**
     * @param {?} element
     * @return {?}
     */
    function (element) {
        var _this = this;
        this.zone.runOutsideAngular(function () { return _this.getElement(element).mCustomScrollbar('destroy'); });
    };
    /**
     * @param {?} element
     * @return {?}
     */
    MalihuScrollbarService.prototype.getElement = /**
     * @param {?} element
     * @return {?}
     */
    function (element) {
        if (typeof element === 'string') {
            return $(element);
        }
        if (typeof element === 'object' && element instanceof HTMLElement) {
            return $(element);
        }
        if (element instanceof jQuery || 'jquery' in Object(element)) {
            return element;
        }
        throw Error("Unsupported element type in MalihuScrollbarService: " + element);
    };
    MalihuScrollbarService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    MalihuScrollbarService.ctorParameters = function () { return [
        { type: NgZone }
    ]; };
    return MalihuScrollbarService;
}());
export { MalihuScrollbarService };
if (false) {
    /** @type {?} */
    MalihuScrollbarService.prototype.zone;
}
